File: org.kevoree.modeling.microframework/src/test/java/org/kevoree/modeling/microframework/test/polynomial/PolynomialTest.java
package org.kevoree.modeling.microframework.test.polynomial;

import org.junit.Test;

import java.util.Random;
import java.util.TreeMap;

/**
 * Created by duke on 10/28/14.
 */
public class PolynomialTest {

    @Test
    public void test() {
        TreeMap<Long, Double> testVal = new TreeMap<Long, Double>();
        Random rand = new Random();
        int degradeFactor = 1;
        double toleratedError = 0.001;
        int maxDegree = 20;
        long starttime;
        long endtime;
        double res;

        PolynomialModel pm = new PolynomialModel(degradeFactor,toleratedError,maxDegree);

        Long l;
        Double d;
        long initTimeStamp=0;
        long finalTimeStamp = 1000000;

        for(long i=initTimeStamp; i<finalTimeStamp;i++){
            l= new Long(i);
            d= new Double (rand.nextDouble());
            testVal.put(l,d);
            pm.feed(l,d);
        }
        pm.finalSave();

        starttime = System.nanoTime();
        StatClass sc = pm.displayStatistics(true);
        endtime = System.nanoTime();
        res=((double)(endtime-starttime))/(1000000);
        System.out.println("Statistic calculated in: "+res+" ms!");

        //Replace by assert
        System.out.println("Max error respected: " +String.valueOf(sc.maxErr<toleratedError));

        starttime = System.nanoTime();
        for(long i=initTimeStamp; i<finalTimeStamp;i+=degradeFactor){
            double val = pm.reconstruct(i);
        }
        endtime = System.nanoTime();
        res=((double)(endtime-starttime))/(1000000);
        System.out.println("Polynomial chain reconstructed in: "+res+" ms!");


        starttime = System.nanoTime();
        for(long i=initTimeStamp; i<finalTimeStamp;i+=degradeFactor){
            double val = pm.fastReconstruct(i);
        }
        endtime = System.nanoTime();
        res=((double)(endtime-starttime))/(1000000);
        System.out.println("Polynomial fast reconstructed in: "+res+" ms!");


        starttime = System.nanoTime();
        for(long i=initTimeStamp; i<finalTimeStamp;i+=degradeFactor){
            double val = testVal.get(testVal.floorKey(i));
        }
        endtime = System.nanoTime();
        res=((double)(endtime-starttime))/(1000000);
        System.out.println("normal chain in: "+res+" ms!");
    }

}

File: org.kevoree.modeling.microframework.typescript/src/test/ts/org/kevoree/modeling/microframework/test/polynomial/PolynomialTest.ts
///<reference path="../../../../../junit/Test.ts"/>
///<reference path="../../../../../../java/util/Random.ts"/>
///<reference path="../../../../../../java/util/TreeMap.ts"/>
// TODO Resolve static imports
///<reference path="../../../../../junit/Assert.ts"/>

class PolynomialTest {

  public test(): void {
    var testVal: TreeMap<number, number> = new TreeMap<number, number>();
    var rand: Random = new Random();
    var degradeFactor: number = 10;
    var toleratedError: number = 0.001;
    var maxDegree: number = 20;
    var starttime: number;
    var endtime: number;
    var res: number;
    var pm: PolynomialModel = new PolynomialModel(degradeFactor, toleratedError, maxDegree);
    var l: number;
    var d: number;
    var initTimeStamp: number = 0;
    var finalTimeStamp: number = 10000000;
    for (var i: number = initTimeStamp; i < finalTimeStamp; i += degradeFactor) {
      l = new number(i);
      d = new number(rand.nextDouble());
      testVal.put(l, d);
      pm.feed(l, d);
    }
    pm.finalSave();
    starttime = System.nanoTime();
    var sc: StatClass = pm.displayStatistics(true);
    endtime = System.nanoTime();
    res = (<number>(endtime - starttime)) / (1000000);
    System.out.println("Statistic calculated in: " + res + " ms!");
    System.out.println("Max error respected: " + String.valueOf(sc.maxErr < toleratedError));
    assertTrue(sc.maxErr < toleratedError);
    starttime = System.nanoTime();
    for (var i: number = initTimeStamp; i < finalTimeStamp; i++) {
      pm.reconstruct(i);
    }
    endtime = System.nanoTime();
    res = (<number>(endtime - starttime)) / (1000000);
    System.out.println("Polynomial chain reconstructed in: " + res + " ms!");
    starttime = System.nanoTime();
    for (var i: number = initTimeStamp; i < finalTimeStamp; i++) {
      pm.fastReconstruct(i);
    }
    endtime = System.nanoTime();
    res = (<number>(endtime - starttime)) / (1000000);
    System.out.println("Polynomial fast reconstructed in: " + res + " ms!");
    starttime = System.nanoTime();
    for (var i: number = initTimeStamp; i < finalTimeStamp; i++) {
      testVal.get(testVal.floorKey(i));
    }
    endtime = System.nanoTime();
    res = (<number>(endtime - starttime)) / (1000000);
    System.out.println("normal chain in: " + res + " ms!");
  }

}

